### ISSUE TITLE
Use correct location for Salt PKI

### ISSUE THREAD

**Issue by eliasp** [saltstack/salt#28070]
Currently, the PKI used by Salt resides by default in `/etc/salt/pki` which is strictly speaking not the correct location - `/etc` is reserved for configuration files.
The correct location would be `/var/lib/salt/pki`.

I'd like to see Salt making use of `/var/lib/salt/pki` as default PKI location in future releases.
For a few major releases it might make sense to check `/etc/salt/pki` for legacy PKI data and warn the user about it.

**Comment by jfindlay** [comment]
@eliasp, thanks for the report.  Is there a specific standard in [FHS](http://www.pathname.com/fhs/pub/fhs-2.3.html) that recommends this?

**Comment by eliasp** [comment]
@jfindlay The "[/var/lib : Variable state information](http://www.pathname.com/fhs/pub/fhs-2.3.html#VARLIBVARIABLESTATEINFORMATION)" section covers this.

Also in the interest of proper, clean and reproducible deployments (see also: [systemd - Factory Reset, Stateless Systems, Reproducible Systems & Verifiable Systems](http://0pointer.net/blog/projects/stateless.html)) using the purpose-based locations becomes even more important.

In a config management/SaltStack driven world, I wouldn't even include `/etc` in a backup as I'd assume it's just configuration data, which is regenerated by SaltStack anyways while I'd backup `/var/lib` as I expect it to contain my application state information.

Not messing with users' expectations will ensure a smooth and predictable experience for sysadmins.

**Comment by jfindlay** [comment]
@eliasp, what is your recommendation for transitioning to the new directory layout?  This change would affect basically every salt user.

**Comment by eliasp** [comment]
I think the following approach would be reasonable:
- set the default for `pki_dir` to `/var/lib/salt/pki`
- add a few lines of code in `salt/config.py` which:
  - checks `/var/lib/salt/pki` for PKI data, if there are none, but there are in `/etc/salt/pki`:
    - falls back to `/etc/salt/pki` for `opts['pki_dir']`
    - logs a deprecation warning and asks the user to move the data

Moving the PKI data could even be done by Salt itself instead of displaying a deprecation warning. This would possibly fit quite well as the whole PKI magic is already managed quite well by Salt without user interference.
The whole fallback-mechanism should only kick in, in case the user didn't explicitely set `pki_dir` on its own.
The fallback to `/etc/salt/pki` could possibly be dropped after 4-5 major releases.

This change would affect (as far as I can see):
**Salt:**
* `conf/master` and `conf/minion` - a simple `sed` will be sufficient here
* `doc/*` - for the most occurences of `/etc/salt/pki`, the same as for `conf/*` applies. There are a few exceptions where the docs refer to the split of the PKI dir into `master` and `minion` in release `0.11.0`
* `pkg/*` - the same here, with the exception of `pkg/salt.postrm` which should keep the statement to remove `/etc/salt/pki/$1` for backwards-compatibility and a statement to remove `/var/lib/salt/pki/$1` should probably be added
* `salt/cloud/deploy/*.sh` - nearly every deploy script does something along the lines of:

    ```
mkdir -p /etc/salt/pki
echo '{{ vm['priv_key'] }}' > /etc/salt/pki/minion.pem
echo '{{ vm['pub_key'] }}' > /etc/salt/pki/minion.pub
```

**Salt Bootstrap:**
* `salt/config.py` currently places `pki_dir` below `salt.syspaths.CONFIG_DIR`
* `salt/syspaths.py` needs to add a new constant like `VARLIB_DIR` or `PERSISTENCE_DIR` to be used in `salt/config.py` instead for setting the default for `pki_dir`.
* `bootstrap-salt.sh` places `_PKI_DIR` relative to `_SALT_ETC_DIR`. Here `_SALT_ETC_DIR` needs to be replaced with a new variable `_SALT_VARLIB_DIR`
* `bootstrap-salt.ps1` creates the directory `C:\salt\conf\pki\minion` - this should be `C:\salt\var\lib\pki\minion` instead. Although on Windows the FHS-like separation doesn't really matter, this consistency with POSIX environments will make the overall handling easier

**Comment by jfindlay** [comment]
@cachedout

**Comment by cachedout** [comment]
Hi Elias,

Thanks for the thorough explanation. I do like the idea of keys located in
/var/lib/salt/pki, but I do have to wonder about the benefits of this
versus the potential headaches for people during a transition. If we did
this, it would have to be _very_ slowly, with a well-documented deprecation
path over a few releases.

At present, @thatch45 is away from the office but when he gets back I would
like to include him in the conversation. Such a big change would require a
much larger discussion, but I'm totally open to the idea if we can figure
out how to do it while keeping the impact of an upgrade to this new path to
a minimum.

**Comment by eliasp** [comment]
I agree. This is nothing to be done in a rush. It's not an imminent problem, just a change to "do things right" and it doesn't matter if it takes 2 or 5 major releases until this migration is done.

**Comment by eliasp** [comment]
Keep this one open…

**Comment by brianholland99** [comment]
FYI - At least the Ubuntu 18.04 salt-master and salt-minion installs from the standard apt-get repos seem to have made the change to /var/lib/salt/pki.  It would be nice to converge to this location and some well-documented notification and roadmap (even deprecation warning) as mentioned above so people can prepare. I either need to have scripts handle both locations or possibly switch back to salt bootstrap to pull in the upstream version.

Things to include when doing changes:
  - Documentation on preseeding keys should change to match new default location.
  - /etc/salt/master and /etc/salt/minion commented out pki_dir settings should show the new default location.

I only found this ticket existed when I changed from salt bootstrap that I had been using in 16.04 to the standard Ubuntu packages now that they had the change that I needed. The packages in the standard Ubuntu repos unfortunately left the commented-out pki_dir pointing to /etc/salt/pki, so took a 'find /' to locate the new location and then search the web to find this ticket.
